src/Orleans.Transactions/State/TransactionalStateFactory.cs

```diff
@@ -0,0 +1,23 @@
+﻿using Microsoft.Extensions.DependencyInjection;
+using Orleans.Runtime;
+using Orleans.Transactions.Abstractions;
+
+namespace Orleans.Transactions
+{
+    public class TransactionalStateFactory : ITransactionalStateFactory
+    {
+        private IGrainActivationContext context;
+
+        public TransactionalStateFactory(IGrainActivationContext context)
+        {
+            this.context = context;
+        }
+
+        public ITransactionalState<TState> Create<TState>(ITransactionalStateConfiguration config) where TState : class, new()
+        {
+            TransactionalState<TState> transactionalState = ActivatorUtilities.CreateInstance<TransactionalState<TState>>(this.context.ActivationServices, config, this.context);
+            transactionalState.Participate(context.ObservableLifecycle);
```

Comment:
Could `TransactionalState<>` subscribe to the lifecycle in its ctor instead? I might have asked a similar question before in another context. Apologies if so.

Reply:
It 'can', it's just unsafe.  If it registers itself with the lifecycle then something goes wrong with the construction (error thrown) an invalid object will be registered in the lifecycle.